### PR TITLE
feat(hummock/meta): Reliable task cancellation via heartbeats + external task cancellation gRPC

### PR DESCRIPTION
**Pull request by jon-chuang**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
Cancel compaction tasks and trigger new ones to its compaction group if no progress is made in `num_blocks_completed` or `num_ssts_uploaded` after a certain timeout.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
RFC: https://github.com/singularity-data/risingwave/issues/4457
Fixes: https://github.com/singularity-data/risingwave/issues/3677, https://github.com/singularity-data/risingwave/issues/4511, https://github.com/risingwavelabs/risingwave/issues/3677

### CODE REVIEW

File: src/storage/src/hummock/compactor.rs
Comment (jon-chuang) [inline src/storage/src/hummock/compactor.rs:972]: remove outdated

**Comment by jon-chuang** [comment]
As pointed out in https://github.com/singularity-data/risingwave/issues/4379#issuecomment-1208772404, we actually need to retry `report_compact_task` in a loop until it succeeds, particularly important in the compaction heartbeat loop as it is a catch all for reporting task status.

It remains to be decided if we use an infinite loop or retry a fixed number of times. I would go with the latter and then try to schedule for the group manually as a robust fail all. If all of the above fails, we report error and advise that LSM tree may be permanently imbalanced.

File: src/meta/src/hummock/compactor_manager.rs
Comment (jon-chuang) [inline src/meta/src/hummock/compactor_manager.rs:176]: TODO: implement this. We should add this method to `Subscribe...Response`

File: src/meta/src/hummock/manager/tests.rs
Comment (zwang28) [inline src/meta/src/hummock/manager/tests.rs:1049]: Time dependent tests can be flaky. But I think it won't be a problem in [deterministic test](https://github.com/singularity-data/risingwave/pull/4476).
Maybe we want an attribute macro that only enables test_hummock_compaction_task_heartbeat in deterministic test?

File: src/rpc_client/Cargo.toml
Comment (zwang28) [inline src/rpc_client/Cargo.toml:28]: We use tokio-stream = "0.1" in other crates.

File: src/meta/src/hummock/manager/mod.rs
Comment (zwang28) [inline src/meta/src/hummock/manager/mod.rs:256]: Reschedule is optional  because eventually another write will trigger it.

File: src/meta/src/hummock/compactor_manager.rs
Comment (zwang28) [inline src/meta/src/hummock/compactor_manager.rs:41]: I think task_heartbeats need to be initialized with existent assigned tasks, during meta node bootstrap.

File: src/meta/src/hummock/manager/mod.rs
Comment (zwang28) [inline src/meta/src/hummock/manager/mod.rs:239]: If we don't consume tasks in get_timed_out_tasks, then the retry here is optional and report_compact_task will succeed in later loop eventually.

Comment (jon-chuang) [reply src/meta/src/hummock/manager/mod.rs:239]: Hmm, we could indeed not consume if not successful. I think that is the better behaviour. And the retry behaviour occurs quite regularly, so perhaps we can get rid of all the retry and fallback logic here.

Comment (jon-chuang) [reply src/meta/src/hummock/manager/mod.rs:239]: I think we are ok with a task not being successfully reported indefinitely? As we consider that quite unlikely? 

The retry behaviour here is at least bounded but the result is still a task stuck indefinitely. So I would say retrying indefinitely is not worse.

File: src/meta/src/hummock/manager/mod.rs
Comment (jon-chuang) [reply src/meta/src/hummock/manager/mod.rs:256]: Well, I think the point is if there are no new writes. But anyway, perhaps it is good enough to retry reporting task status indefinitely and not have a restart fallback as it's quite unlikely for two errors to cooincide (can't report, and no new writes).

File: src/meta/src/hummock/manager/mod.rs
Comment (jon-chuang) [inline src/meta/src/hummock/manager/mod.rs:939]: Done, remove outdated

File: src/meta/src/hummock/manager/tests.rs
Comment (jon-chuang) [reply src/meta/src/hummock/manager/tests.rs:1049]: Maybe `#[cfg(madsim)]` @wangrunji0408 ?

File: src/meta/src/hummock/compactor_manager.rs
Comment (zwang28) [reply src/meta/src/hummock/compactor_manager.rs:41]: Otherwise if a problematic compactor no longer sends compaction heartbeat, meanwhile the meta is rebooted, then the tasks will never show in task_heartbeats.

File: src/storage/src/hummock/compactor.rs
Comment (jon-chuang) [inline src/storage/src/hummock/compactor.rs:991]: Todo: add back \`consume_stream

File: src/storage/src/hummock/compactor.rs
Comment (jon-chuang) [inline src/storage/src/hummock/compactor.rs:1007]: Continue \`consume_stream

File: src/storage/src/hummock/sstable/multi_builder.rs
Comment (jon-chuang) [inline src/storage/src/hummock/sstable/multi_builder.rs:177]: It's ok to unwrap guard.

File: src/storage/src/hummock/sstable/multi_builder.rs
Comment (jon-chuang) [inline src/storage/src/hummock/sstable/multi_builder.rs:142]: We can use the cloned task_progress instead of new var.

File: src/storage/src/hummock/compactor.rs
Comment (jon-chuang) [inline src/storage/src/hummock/compactor.rs:939]: Maybe this trace is not necessary. The receiver will already report task cancelled externally.

File: src/storage/src/hummock/compactor.rs
Comment (jon-chuang) [inline src/storage/src/hummock/compactor.rs:617]: Add task_id to message. And maybe compact task info?

File: src/meta/src/hummock/manager/tests.rs
Comment (wangrunji0408) [reply src/meta/src/hummock/manager/tests.rs:1049]: Yes, just add it to this test.

File: src/meta/src/hummock/manager/mod.rs
Comment (jon-chuang) [inline src/meta/src/hummock/manager/mod.rs:225]: Updated outdated and confusing comments

File: src/meta/src/hummock/compactor_manager.rs
Comment (jon-chuang) [reply src/meta/src/hummock/compactor_manager.rs:41]: Yes, we should persist any tasks to meta, and remove them too. But the overhead (per-task) could be quite high.

See https://github.com/singularity-data/risingwave/blob/fb24e62fbdbd9006b61bd7971997ac72f2e5a56e/src/meta/src/cluster/mod.rs#L342

Comment (jon-chuang) [reply src/meta/src/hummock/compactor_manager.rs:41]: Just as a note, the correct order of behaviour is to first add heartbeat (persist) then send the task creation request. Even if request fails, it's still fine as heartbeat timeout will elapse eventually.

**Comment by zwang28** [comment]
There is one more refactor we may consider in this PR, or later PR:
Currently `assign_compaction_task` executes both task assignment and task send. It was originally written this way to ease error handling. But it's improper because 1. it requires compaction lock while task send is slow. 2. task is sent before assigned (see `assign_compaction_task`).

https://github.com/singularity-data/risingwave/blob/bb68c57903eb896c9d65e043f72d22923dbd9603/src/meta/src/hummock/compaction_scheduler.rs#L187

File: src/meta/src/hummock/compactor_manager.rs
Comment (zwang28) [reply src/meta/src/hummock/compactor_manager.rs:41]: > Yes, we should persist any tasks to meta store and remove them when completed too

We've already persisted it in 
https://github.com/singularity-data/risingwave/blob/ef5de060f4d317866986df3f7dff1c21af8b04f7/src/meta/src/hummock/manager/compaction.rs#L25

It requires a meta store write per op.

**Comment by jon-chuang** [comment]
That sounds quite orthogonal so perhaps I'll look into it in follow up PR. I'll create an issue for it.

File: proto/hummock.proto
Comment (hzxa21) [inline proto/hummock.proto:340]: I think we can just make SubscribeCompactTasks rpc to be a bidi-streaming rpc and use this streaming rpc as the only compaction related communication channel so that we don't need to create a dedicated unary rpc for CompactTaskProgress reporting.

Comment (zwang28) [reply proto/hummock.proto:340]: I don't quite get the benefit of unifying all RPCs into a stream, because we still have to dispatch messages based on their type if using stream?

Comment (hzxa21) [reply proto/hummock.proto:340]: > we still have to dispatch messages based on their type if using stream

True. The dispatch logic is simple though. IMO, the main benefit is to simplify failure handling and retry:
- We have one unify logic to handle rpc failure, which is to re-create the stream. We don't need to consider the case when we have one rpc fails but the other rpc still works.
- We don't need to worry about timeout / retry since the streaming rpc has its own mechanism to ensure delivery.

The side benefits are:
- Reduce the number of messages transmitted because we don't need to explicitly ack each request of ReportCompactionTask / ReportCompactionTaskProgress.
- Reduce overhead of rpc creation/destruction. In gRPC, some application-level metadata needs to be transmitted between two ends to initiate/finish a rpc and we can avoid these work by have one streaming rpc instead of many unary rpcs.

(technially these side benefits are the benefits of using streaming rpc, not unifying all rpcs into one streaming rpc)

**Comment by hzxa21** [comment]
Any updates?

File: proto/hummock.proto
Comment (jon-chuang) [reply proto/hummock.proto:341]: I tried implementing as a stream, it was quite troublesome to manage incoming streams. We need to add a task to consume the stream rather than allow new RPC to trigger a task. 

Anyway, maybe this is incremental work that can be reserved for next PR if it is deemed necessary. 

Basically, we should just have a global stream, and each compactor added to meta will pipe the input stream into the global stream to be consumed by the global "update compactor progress" task.

**Comment by jon-chuang** [comment]
Sorry. I've merged main today. I'll try to address all the review comments by tomorrow morning.

File: proto/hummock.proto
Comment (jon-chuang) [reply proto/hummock.proto:345]: Anw, this fits the mpsc pattern, we can actually clone the tx for each compactor. I'm just not sure how we'd manage the lifetime.

File: src/meta/src/hummock/compactor_manager.rs
Comment (jon-chuang) [reply src/meta/src/hummock/compactor_manager.rs:44]: Then in the current model we may also need to bootstrap the compactor connections. But this may not be feasible. It may be better to bootstrap the existing tasks from any connection that does survive.

Comment (jon-chuang) [reply src/meta/src/hummock/compactor_manager.rs:44]: Also, maybe decoupling the heartbeat insertion to time of assignment rather than time of sending may be beneficial.

Comment (jon-chuang) [reply src/meta/src/hummock/compactor_manager.rs:44]: One issue with this is that to ensure state consistency, we need to perform hearbeat initiation as part of the task assignment transaction. Else, we may assign a task which for some reason never has its hearbeat initiated, thus leaving it stuck in limbo.

Comment (jon-chuang) [reply src/meta/src/hummock/compactor_manager.rs:44]: Actually, if we simply assume that heartbeat initiation is reliable, this is not a problem.

File: src/meta/src/hummock/compaction_scheduler.rs
Comment (jon-chuang) [inline src/meta/src/hummock/compaction_scheduler.rs:110]: this is not necessary...anw, there is no need to break 'compaction_trigger, there is only one loop so break is fine.

**Comment by jon-chuang** [comment]
I've made the suggestion changes. In addition, the unit test runs and passes with RUSTFLAGS="--cfg madsim"

File: src/meta/src/hummock/compactor_manager.rs
Comment (zwang28) [inline src/meta/src/hummock/compactor_manager.rs:211]: It's possible two concurrent `add_compactor` both `self.remove_compactor` and then `guard.compactors.push`, which causes duplicate items in `compactors`?

File: src/storage/src/hummock/compactor/mod.rs
Comment (zwang28) [inline src/storage/src/hummock/compactor/mod.rs:264]: Could you please explain how many benefits external cancellation will bring, with its implementation complexity introduced?

**Review by zwang28 — COMMENTED** [review]
Rest LGTM, cool.

File: src/storage/src/hummock/compactor/mod.rs
Comment (jon-chuang) [reply src/storage/src/hummock/compactor/mod.rs:249]: There's only one conceivable benefit, which is early stopping of tasks on compactor side if meta side deems task dead.

Comment (jon-chuang) [reply src/storage/src/hummock/compactor/mod.rs:249]: I believe the change is small compared to the rest of the PR (which is reliable heartbeats).

File: src/meta/src/hummock/compactor_manager.rs
Comment (jon-chuang) [reply src/meta/src/hummock/compactor_manager.rs:211]: Yes, perhaps I will make the operation atomic under a single guard to avoid concurrency issues.

File: src/meta/src/rpc/service/hummock_service.rs
Comment (Little-Wallace) [inline src/meta/src/rpc/service/hummock_service.rs:207]: Suggest streaming interface

File: src/storage/src/hummock/compactor/mod.rs
Comment (jon-chuang) [reply src/storage/src/hummock/compactor/mod.rs:249]: In hindsight, it could have been seperated into a separate PR for simplicity's sake.

File: src/meta/src/rpc/service/hummock_service.rs
Comment (jon-chuang) [reply src/meta/src/rpc/service/hummock_service.rs:207]: I've proposed to implement it in next PR as this PR is quite long and its not so trivial to do the streaming as mentioned in this thread: https://github.com/singularity-data/risingwave/pull/4496#discussion_r944177447

**Comment by jon-chuang** [comment]
My apologies, I fell sick in the latter half of last week. I've cleaned up some remaining issues.

File: src/storage/src/hummock/sstable/multi_builder.rs
Comment (Little-Wallace) [inline src/storage/src/hummock/sstable/multi_builder.rs:201]: Why report blocks uploaded? Meta can not judge when this progress tracker could finish by the number of uploaded blocks.

Comment (jon-chuang) [reply src/storage/src/hummock/sstable/multi_builder.rs:201]: > Meta can not judge when this progress tracker could finish by the number of uploaded blocks. 

Could you clarify what you mean?

We use either increment uploaded blocks or increment in sealed blocks as a measure of progress. In the existing code path, both are required for the entire task to be reported as complete.

**Review by zwang28 — APPROVED** [review]
LGTM.
@Li0k Would you please take a look at this PR?

File: src/storage/src/hummock/sstable/multi_builder.rs
Comment (jon-chuang) [reply src/storage/src/hummock/sstable/multi_builder.rs:201]: Oh, I see, I think I made a mistake, this only measures progress at granularity of table, not block.

Comment (jon-chuang) [reply src/storage/src/hummock/sstable/multi_builder.rs:201]: I will change to `inc_tables_uploaded`. Further, I will push progress tracker into each `SstableBuilder`, so that `inc_blocks_built` will be called upon every `build_block`

File: risedev.yml
Comment (zwang28) [inline risedev.yml:29]: Is this change committed by mistake?

Comment (jon-chuang) [reply risedev.yml:29]: Yes :)

**Comment by jon-chuang** [comment]
Hi @Li0k would you please take a look? :)

**Comment by Li0k** [comment]
> Hi @Li0k would you please take a look? :)

Sorry, I will finish the review this two days

File: src/meta/src/hummock/compactor_manager.rs
Comment (jon-chuang) [reply src/meta/src/hummock/compactor_manager.rs:212]: Yes, I'm not sure where the context_id is generated, let me check to see if this is possible.

File: src/meta/src/hummock/compaction_scheduler.rs
Comment (jon-chuang) [inline src/meta/src/hummock/compaction_scheduler.rs:113]: should be "Shutdown compactor scheduler"

File: src/meta/src/hummock/compactor_manager.rs
Comment (jon-chuang) [reply src/meta/src/hummock/compactor_manager.rs:212]: So, it is not possible for two nodes to have the same worker ID if they run our code. However, it is entirely possible for a single worker to try to register twice, the second request gets overwritten by the first, so meta holds connection to first, which is dropped, and compactor holds connection to second, which meta drops.

However, even in this scenario, we already have the necessary retry logic. The compactor will receive a None message, indicating it needs to retry subscribe_compact_tasks. If it suceeds this time, meta will replace the dropped compator connection with new connection.

Anyway, we can add a sync point test to test this particular scenario (and others).

File: src/meta/src/hummock/manager/mod.rs
Comment (zwang28) [inline src/meta/src/hummock/manager/mod.rs:232]: Should use cancel_compact_task instead of report_compact_task.

report_compact_task(context_id, &task) will fail if context_id is not a valid member of the cluster.

Comment (zwang28) [inline src/meta/src/hummock/manager/mod.rs:232]: Should use cancel_compact_task instead of report_compact_task.

report_compact_task(context_id, &task) will fail if context_id is not a valid member of the cluster.

File: src/meta/src/hummock/manager/mod.rs
Comment (zwang28) [inline src/meta/src/hummock/manager/mod.rs:947]: After using cancel_compact_task in line 232, this Some(context_id) = context_id will never be true. I think cancel_compact_task can be called after line 232 succeeds.

Comment (jon-chuang) [reply src/meta/src/hummock/manager/mod.rs:947]: I think we should rename cancel_compact_task, it should be cancel_unassigned_compact_task. 

Btw, there is a function that cancels ALL unassigned tasks I believe.

Comment (jon-chuang) [reply src/meta/src/hummock/manager/mod.rs:947]: Anw the main thing is we need to simplify the interface, there is a lot of implicit assumptions about it's usage and the usage of report_compact_task atm

Comment (zwang28) [reply src/meta/src/hummock/manager/mod.rs:947]: > I think we should rename cancel_compact_task, it should be cancel_unassigned_compact_task.

In https://github.com/singularity-data/risingwave/pull/4932 I mean to make cancel_compact_task's semantic "cancel task anyway".

**Comment by zwang28** [comment]
Note that this PR doesn't handle cancellation of picked but unassigned task. 

As a result, we must ensure an unassigned task is eventually either assigned, or cancelled during runtime. (Unassigned tasks are also cancelled during meta bootstrap, but that's the last guard).
So here is one mistake I need to fix: this method tries to cancel task only once, rather than ensures its success
https://github.com/singularity-data/risingwave/blob/9cd73e12ba1aa39992c5a619ad2f72eca258e8e4/src/meta/src/hummock/manager/mod.rs#L1388

As a comparison, this code will ensure an unassigned task is eventually either assigned, or cancelled:
https://github.com/singularity-data/risingwave/blob/9cd73e12ba1aa39992c5a619ad2f72eca258e8e4/src/meta/src/hummock/compaction_scheduler.rs#L153

File: src/meta/src/hummock/manager/mod.rs
Comment (jon-chuang) [reply src/meta/src/hummock/manager/mod.rs:947]: I think we should require report_compact_task to provide the context_id, only cancel will use None for report_compact_task_impl

File: src/meta/src/rpc/service/hummock_service.rs
Comment (Li0k) [reply src/meta/src/rpc/service/hummock_service.rs:208]: Agree with `streaming RPC`, shall we add some `TODO` comments for this so that we can easily locate it later.

**Review by Li0k — APPROVED** [review]
Rest LGTM, great job !!! Thanks for this PR

File: src/storage/src/hummock/sstable/multi_builder.rs
Comment (Li0k) [inline src/storage/src/hummock/sstable/multi_builder.rs:210]: I'm a little concerned about lock contention, even though we will only operate once per sst.

Is there a test here？ like benchmark or mirco_bench (like `bench_compactor`)

File: src/meta/src/hummock/manager/mod.rs
Comment (zwang28) [reply src/meta/src/hummock/manager/mod.rs:947]: `if let Some(assignee_context_id) = assignee_context_id {` ?

**Comment by hzxa21** [comment]
@zwang28 @Li0k @jon-chuang Shall we prioritize merging this PR?

**Comment by zwang28** [comment]
> Shall we prioritize merging this PR

Need resolve these for correctness @jon-chuang 
https://github.com/risingwavelabs/risingwave/pull/4496#discussion_r956209913
https://github.com/risingwavelabs/risingwave/pull/4496#discussion_r956837883

**Comment by Little-Wallace** [comment]
Seems a lot conflict to be resolved.

**Comment by Gun9niR** [comment]
> Seems a lot conflict to be resolved.

It's ok to prioritize this PR first.

File: src/meta/src/hummock/manager/mod.rs
Comment (jon-chuang) [reply src/meta/src/hummock/manager/mod.rs:947]: Makes sense, we always use the actual state. I will make the cluster membership thread purge remaining heartbeats forcefully if 

I will add a test that when a context is removed, the task heartbeats are all removed.

File: src/storage/src/hummock/sstable/multi_builder.rs
Comment (jon-chuang) [reply src/storage/src/hummock/sstable/multi_builder.rs:210]: Actually, there is potential contention across splits.

I have not done any bench or test so far.

File: src/meta/src/hummock/manager/mod.rs
Comment (jon-chuang) [reply src/meta/src/hummock/manager/mod.rs:947]: So the behaviour is to remove task heartbeat IFF reported and assignee is not None OR context is removed.

File: src/meta/src/hummock/manager/mod.rs
Comment (jon-chuang) [reply src/meta/src/hummock/manager/mod.rs:226]: I've added TODO for metrics an tracing

File: src/storage/src/hummock/sstable/multi_builder.rs
Comment (jon-chuang) [reply src/storage/src/hummock/sstable/multi_builder.rs:210]: anw, I don't think it will be so bad. block size is 128MB meaning that one unit of execution will ask for lock relatively infrequently. While it may cause context switch of underlying threads if blocked on the lock, the underlying operation is very fast (it is just an increment of an integer) so I think the chance of that is low.

Comment (jon-chuang) [reply src/storage/src/hummock/sstable/multi_builder.rs]: I've changed it to increment a block counter instead of sst counter now, this makes things worse. In fact, to avoid lock contention, I believe it is sufficient to simply increment a counter for every SST produced as before. Anw, we do not have a very low latency requirement for the identification of stalled tasks and we currently already choose a 60 second task heartbeat timeout.

**Comment by zwang28** [comment]
Compaction is triggered by LSM write, so currently it's possible no compaction is triggered when a cluster is (re)started and no write is committed.
Let's fix this in another PR.

**Comment by jon-chuang** [comment]
> Note that this PR doesn't handle cancellation of picked but unassigned task.

Is this fixed?

> Compaction is triggered by LSM write, so currently it's possible no compaction is triggered when a cluster is (re)started and no write is committed.
> Let's fix this in another PR.

Indeed, we need to trigger compaction for all groups whenever we restart the cluster. Since this is a noop if a particular compaction group will not produce a task, there is no cost to doing this.

**Comment by zwang28** [comment]
> Indeed, we need to trigger compaction for all groups whenever we restart the cluster.

And there are more cases that requires an explicit trigger, e.g. a compactor is recovered.

> (in fact, there is no cost to periodically checking in a tokio task for new compaction tasks for all compaction groups rather than to have a reactive approach as we do now... Perhaps this is a better solution as it can help us make progress on compaction even if we've made a mistake or there is a fault somewhere in our triggering of new tasks?)

The cost of periodically checking is [this RW lock](https://github.com/risingwavelabs/risingwave/blob/047802d2048875a9369a6d60ea7a2c774b6a51de/src/meta/src/hummock/manager/mod.rs#L80). That is, checking task may block reporting task, as checking do takes time.

Yes a periodical checking is needed. I think the benefits of current approach is a new task can be scheduled right after LSM write, rather than waiting for certain time interval.

**Comment by jon-chuang** [comment]
> The cost of periodically checking is [this RW lock](https://github.com/risingwavelabs/risingwave/blob/047802d2048875a9369a6d60ea7a2c774b6a51de/src/meta/src/hummock/manager/mod.rs#L80).
> I think we can periodically send signal to compaction scheduler for all compaction groups.

I see. I think the cost is low if we check relatively infrequently. Say, once every 5 seconds or once every 30 seconds.

**Comment by zwang28** [comment]
I think this PR is ready for merge now? @jon-chuang

**Comment by jon-chuang** [comment]
I've tested this PR in e2e setting by commenting out compact_done.

When commented, we get our desired tracing messages

e.g.

```
2022-09-05T07:41:25.189742Z  INFO risingwave_meta::hummock::manager: Task with task_id 42 with context_id 1 has expired due to lack of visible progress
2022-09-05T07:41:25.189935Z  INFO risingwave_meta::hummock::manager: CancelTask operation for task_id 42 has been sent to node with context_id 1
```

After merging, I'll test the PR again to ensure that long running tasks will not trigger task expiry as long as they make progress (my tpch-bench is not working properly atm).

File: src/storage/src/hummock/sstable/multi_builder.rs
Comment (jon-chuang) [reply src/storage/src/hummock/sstable/multi_builder.rs]: Actually, now that I think about it, implementing as atomic may be more efficient. I'll create an issue for this.

**Pull request merged**
